### ISSUE TITLE
Stereo recording

### ISSUE THREAD

**Issue by hiroshihorie** [livekit/client-sdk-swift#161]
Automatically enable stereo recording when device supports it.

**Comment by pblazej** [comment]
@hiroshihorie I started playing with that, however:
- it will require switching to the bottom/back mic 🎤 
- for some reason, the supported modes are different from the Apple [example](https://developer.apple.com/documentation/avfaudio/capturing_stereo_audio_from_built-in_microphones)
```
(lldb) po dataSources
▿ 3 elements
  - 0 : <AVAudioSessionDataSourceDescription: 0x302658f30, ID = 1835216945; name = Bottom>
  - 1 : <AVAudioSessionDataSourceDescription: 0x302658f40, ID = 1835216946; name = Front>
  - 2 : <AVAudioSessionDataSourceDescription: 0x302658f60, ID = 1835216947; name = Back>
(lldb) po newDataSource
<AVAudioSessionDataSourceDescription: 0x302658f60, ID = 1835216947; name = Back>
(lldb) po newDataSource.supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 3 elements
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Omnidirectional
    ▿ 1 : AVAudioSessionPolarPattern
      - _rawValue : Subcardioid
    ▿ 2 : AVAudioSessionPolarPattern
      - _rawValue : Stereo
(lldb) po dataSources[0].supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 1 element
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Omnidirectional
(lldb) po dataSources[1].supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 4 elements
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Omnidirectional
    ▿ 1 : AVAudioSessionPolarPattern
      - _rawValue : Cardioid
    ▿ 2 : AVAudioSessionPolarPattern
      - _rawValue : Cardioid
    ▿ 3 : AVAudioSessionPolarPattern
      - _rawValue : Stereo
(lldb) po dataSources[2].supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 3 elements
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Omnidirectional
    ▿ 1 : AVAudioSessionPolarPattern
      - _rawValue : Subcardioid
    ▿ 2 : AVAudioSessionPolarPattern
      - _rawValue : Stereo
```
whereas for the SDK:
```
(lldb) po session.inputDataSources
▿ Optional<Array<AVAudioSessionDataSourceDescription>>
  ▿ some : 2 elements
    - 0 : <AVAudioSessionDataSourceDescription: 0x3028ea5b0, ID = 1835216946; name = Front>
    - 1 : <AVAudioSessionDataSourceDescription: 0x3028e90d0, ID = 1835216947; name = Back>
(lldb) po session.inputDataSources![0].supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 2 elements
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Cardioid
    ▿ 1 : AVAudioSessionPolarPattern
      - _rawValue : Cardioid
(lldb) po session.inputDataSources![1].supportedPolarPatterns
▿ Optional<Array<AVAudioSessionPolarPattern>>
  ▿ some : 1 element
    ▿ 0 : AVAudioSessionPolarPattern
      - _rawValue : Subcardioid
```
the session mode and options are pretty much the same 🤷

**Comment by gulzardev** [comment]
Is there a way to enable stereo even if the mic doesn’t support it?  @pblazej 
From `LKRTCAudioDeviceModule` maybe ?

The use case is capturing app audio from a broadcast extension. However, it downmixes to a mono channel. From the broadcast extension, I receive stereo audio. I tried modifying the SDP, but it didn’t work.

**Comment by pblazej** [comment]
@gulzardev thank you for reporting! I'll try to reproduce that with brodcast 🤞

**Comment by pblazej** [comment]
@hiroshihorie I tried to revisit this topic:
- stereo capturing is probably not worth the effort given the limitations of noise cancellation etc.
- stereo broadcast, sound effects, music streaming definitely is worth exploring

We discussed the idea of "bypassing ADM" completely, I made a quick and dirty experiment [here](https://github.com/webrtc-sdk/webrtc/tree/blaze/stereo-broadcast) and [there](https://github.com/livekit/client-sdk-swift/tree/blaze/stereo-broadcast). The stereo playout is not working yet, probably there are some missing pieces.

My idea is that in case we decide to go this way, we can:
- simplify a lot of ADM-coupled internals like `AudioManager.shared.mixer.capture(appAudio:)` in manual mode vs just push
- create a better abstraction for the upcoming sound effects API
- create a better abstraction from totally custom use cases, like the internet radio, "bring your own capturer" kind of API

TL;DR I think the decoupling itself brings a lot of added value here.
